### DISCUSSION TITLE
Server Components

### DISCUSSION THREAD

**Discussion by shakyShane** [preactjs/preact#2879] [General]
This is going to be a bigger impact than Hooks were (IMO)

Professionally I adopted Preact on SSR'd sites for 3 core reasons:

- size
- lazy hydration (without losing context)
- full React-ecosystem support through preact/compat

I don't want to lose these by moving from Preact, but this announcement throws a spanner in the works as the examples are pretty mind-blowing.

What are everyone's thoughts here? I imagine this is FAR from a simple compat layer (given the intermediate format, streaming support etc)

**Comment by marvinhagemeister** [comment]
Extracting static tress is something that's been on our mind for a while so we are very interested in that. The work in `wmr` with `preact-iso` is a first step towards that as it allows "islands" (i.e. sub-trees) to be progressively hydrated. The next step is to leverage compiler magic to extract static trees.

To [quote Jason](https://mobile.twitter.com/_developit/status/1341134089802412032):

> Discussion is ongoing, but philosophically Preact is more aligned with delivering progressively-enhanced HTML. We've been working on some stuff, but the timeline is currently late 2021.

Rest assured that we'll keep pushing the boundaries for smaller sites and quicker boot up times.

**Comment by developit** [comment]
Indeed. The result (HTML with islands of interactivity) is something we've already been executing around, but the specific technical approach React is using is not in line with Preact's goals. Preact aims to provide a platform-oriented take on VDOM, and as such we stick to using HTML to describe documents. 

It would certainly be possible to build out the wire protocol and client hydration similar to what they showed, and I'm sure folks will do this atop Preact as add-ons. For now though, we have other avenues to compiling applications to progressively-enhanced HTML that work with static hosting and prerendering too.

**Reply by deleonio** [reply]
Hello, I ask me, why we go back to server?

What about permanent data transfer and offline capability?

**Reply by gu-stav** [reply]
@deleonio My main take away from the react presentation was, the ability to offload heavy dependencies and computations to the server, so that that client only receives the rendered html to which you then can add interactivity again. Reduces bundle sizes and speeds up larger applications due to the partial hydration.

If you need a heavy computation or certain dependency offline, you can still do that on the client. It just offers a way to ship less to clients ...

**Reply by developit** [reply]
Partial hydration doesn't require server components, it requires server-side rendering or prerendering (which we already support). Offloading heavy computation to the server is anything but a new concept - folks doing API-first web application development have been doing this using REST/GraphQL/SOAP/etc for decades. There are numerous technical and nontechnical benefits to building an API for computationally expensive operations - many businesses have even monetized the APIs they developed for their own use as entire secondary product lines.

**Reply by srdjan** [reply]
isn't it cool though, to have JSX used as a DSL for the server components. Elegant, more uniform code base for apps that use React. Also, the underlying streaming format should (could?) be more efficient than plain json/xml/etc. If the team/app is already locked in React, this could work nice for them...

**Comment by redbar0n** [comment]
@marvinhagemeister is the JSX plugin you made that compiles JSX to HTML still the favored approach?

https://twitter.com/marvinhagemeist/status/1370041302935560194?s=21&t=80LTIJFd9Pq3eFROO6OKXA

**Comment by karlhorky** [comment]
Would be great to get an update here so Preact also allows async server-only components, now that

- Next 13 app/ directory is out with pretty good React Server Components DX
- Nuxt also released experimental Server Components https://github.com/nuxt/framework/pull/9972

**Reply by phiberber** [reply]
Is it just me or Preact is looking a little abandoned? The v11 Roadmap has little to no update, so does this.

**Reply by karlhorky** [reply]
Seems like there's some hints at movement on Preact Server Components from @marvinhagemeister over here:

<img width="590" alt="Screenshot 2023-07-10 at 09 35 17" src="https://github.com/preactjs/preact/assets/1935696/9d35ee42-d9a8-4506-9fc9-6e8cfcf31c88">


- https://twitter.com/karlhorky/status/1676681707615920129

**Reply by marvinhagemeister** [reply]
Preact isn't abandoned, quite the opposite. Server components are already available in [Fresh](https://fresh.deno.dev/) (=meta framework for Deno) since June 2022 and I'm currently exploring async server components on top of that (see: https://twitter.com/marvinhagemeist/status/1676535616563970048 ). Whilst server components are a no brainer, async ones are not as they have very real tradeoffs.

V11 has been pushed back mostly because we managed to backport lots of the things into v10. I know that might not be as exciting as a splashy new major release, but it makes it much easier for bigger companies to adopt Preact. There are several ones of those moving to Preact in the recent months, which favour stability and ease of upgrading over everything else. All companies we cannot publicly talk about unless they share it on their terms, which is why there isn't much news around that.

Then there is also the work in the ecosystem itself with [Preact Signals](https://github.com/preactjs/signals/) being released not too long ago. All this work doesn't show up in this repository, but it's work that expands the Preact world and usage substantially.

**Reply by phiberber** [reply]
I'm afraid features are going to be delivered to Fresh and take a long time to be pushed to Preact. I've seen you guys have been exploring with Async Server Components but seems there's no implementation of it outside of Fresh currently.

I didn't know you guys were adding the v11 features into v10. That's really good to know and brings hope and I find it equally exciting actually. 

It's also good to know there's big companies "backing" Preact. Thanks for the friendly response!

**Reply by marvinhagemeister** [reply]
> I'm afraid features are going to be delivered to Fresh and take a long time to be pushed to Preact. I've seen you guys have been exploring with Async Server Components but seems there's no implementation of it outside of Fresh currently.

Nah, that's not true. Just today I shipped support for error boundaries in [preact-render-to-string](https://github.com/preactjs/preact-render-to-string/releases/tag/v6.2.0) because I need that for Fresh. Shipping that in `preact-render-to-string` means that any Preact user can benefit from this even before Fresh users do.

There might be some more complex and more meta-framework-like features like async server components that will land in Fresh first. But that's actually a good thing because it allows us to find all the edge cases before upstreaming a more robust implementation to Preact itself.

Then there is already another big feature planned for Preact apps that will ship soon that started its life as an exploration in Fresh. Preact users will be able to use that at the exact same time as Fresh users as the new feature isn't tied to Fresh at all.

**Reply by RayyanNafees** [reply]
Can we use Preact Async Components ?
